Nullable Reference Types: Compiler should not report CS8614 for a nullable value typed argument when the implemented interface lies within another assembly

**Version Used**: 05/14/18 Nullable Reference Types Preview (`csc` reports `2.8.0.62830 (e595ee27)`) with Visual Studio 15.7.1

**Demonstration Code**:

I've created [PathogenPlayground/TestCs8614](https://github.com/PathogenPlayground/TestCs8614) to demonstrate the issue.

**Steps to Reproduce**:

1. Define an interface in a class library with a method `void Test(int? nullableInt)`
2. Create a separate assembly that references the previous class library.
3. Create the same interface in the new assembly.
4. Create implementations for each interface.

**Expected Behavior**:

No warnings are reported.

**Actual Behavior**:

You receive the following warning, but only for the implementation that uses the interface from the other assembly:

```
warning CS8614: Nullability of reference types in type of parameter 'nullableInt' doesn't match implicitly implemented member 'void 
```

----

The readme on [PathogenPlayground/TestCs8614](https://github.com/PathogenPlayground/TestCs8614) has some additional details about how things are set up in the demonstration.

--- a/src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs
@@ -631,6 +631,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
             {
                 Debug.Assert((object)typeSymbol != null);
                 Debug.Assert(!customModifiers.IsDefault);
+                Debug.Assert(!typeSymbol.IsNullableType() || isNullable == true);
                 _typeSymbol = typeSymbol;
                 _isNullable = isNullable;
                 _customModifiers = customModifiers;
@@ -665,7 +666,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
 
             public override TypeSymbolWithAnnotations AsNotNullableReferenceType()
             {
-                return _isNullable == false ?
+                return _isNullable == false || _typeSymbol.IsNullableType() ?
                     this :
                     new NonLazyType(_typeSymbol, isNullable: false, _customModifiers);
             }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs
@@ -32409,7 +32409,7 @@ partial class B<T> where T : A? { }";
 
         [WorkItem(27008, "https://github.com/dotnet/roslyn/issues/27008")]
         [Fact]
-        public void OverriddenMethodNullableValueTypeParameter()
+        public void OverriddenMethodNullableValueTypeParameter_01()
         {
             var source0 =
 @"public abstract class A
@@ -32422,6 +32422,29 @@ partial class B<T> where T : A? { }";
 @"class B : A
 {
     public override void F(int? i) { }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8, references: new[] { ref0 });
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void OverriddenMethodNullableValueTypeParameter_02()
+        {
+            var source0 =
+@"public abstract class A<T> where T : struct
+{
+    public abstract void F(T? t);
+}";
+            var comp0 = CreateCompilation(source0, parseOptions: TestOptions.Regular8);
+            var ref0 = comp0.EmitToImageReference();
+            var source =
+@"class B1<T> : A<T> where T : struct
+{
+    public override void F(T? t) { }
+}
+class B2 : A<int>
+{
+    public override void F(int? t) { }
 }";
             var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8, references: new[] { ref0 });
             comp.VerifyDiagnostics();

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs
@@ -32406,5 +32406,25 @@ partial class B<T> where T : A? { }";
                 //         (z2 = x2).ToString();
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "z2 = x2").WithLocation(12, 10));
         }
+
+        [WorkItem(27008, "https://github.com/dotnet/roslyn/issues/27008")]
+        [Fact]
+        public void OverriddenMethodNullableValueTypeParameter()
+        {
+            var source0 =
+@"public abstract class A
+{
+    public abstract void F(int? i);
+}";
+            var comp0 = CreateCompilation(source0, parseOptions: TestOptions.Regular8);
+            var ref0 = comp0.EmitToImageReference();
+            var source =
+@"class B : A
+{
+    public override void F(int? i) { }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8, references: new[] { ref0 });
+            comp.VerifyDiagnostics();
+        }
     }
 }

